cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.metadata;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.proto.PreferencesDetail;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.EntityId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import sun.net.www.protocol.http.HttpURLConnection;
+
+import java.io.IOException;
+
+/**
+ * Fetch preferences via REST API calls (using internal endpoint {@code INTERNAL_API_VERSION_3})
+ */
+public class RemotePreferencesFetcherInternal implements AbstractPreferencesFetcher {
+  private static final Gson GSON = new Gson();
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemotePreferencesFetcherInternal(final DiscoveryServiceClient discoveryClient) {
+    this.remoteClient = new RemoteClient(
+      discoveryClient, Constants.Service.APP_FABRIC_HTTP,
+      new DefaultHttpRequestConfig(false), Constants.Gateway.INTERNAL_API_VERSION_3);
+  }
+
+  /**
+   * Get preferences for the given identify
+   */
+  public PreferencesDetail get(EntityId entityId, boolean resolved) throws IOException, NotFoundException {
+    HttpResponse httpResponse;
+    String url = getPreferencesURI(entityId, resolved);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(entityId);
+    }
+    return GSON.fromJson(httpResponse.getResponseBodyAsString(), new TypeToken<PreferencesDetail>() {
```

Comment:
type can be a private static final constant

Reply:
Can you clarify which line of code are you referring to?

Reply:
this line, the `new TypeToken`

Reply:
Oh.. Already fixed